### DISCUSSION TITLE
Possible false positive for Literal and dict[str, int|float]?

### DISCUSSION THREAD

**Discussion by sindzicat** [microsoft/pyright#7201] [Q&A]
Hello! According to [PEP 586](https://peps.python.org/pep-0586/#using-non-literals-in-literal-contexts) the following is type safe:

>```python
>def expects_str(x: str) -> None: ...
>var: Literal["foo"] = "foo"
>
># Legal: Literal["foo"] is a subtype of str
>expects_str(var)
>```

So `Literal["foo"]` is a subtype of `str`. Using this fact I try the following:

```python
UnitLengthLiterals = Literal['m', 'dm', 'cm', 'mm', 'mkm', 'km']

s: dict[UnitLengthLiterals, int|float] = {
        'm' : 1,
        'dm' : 0.1,
        'cm' : 0.01,
        'mm' : 0.001,
        'mkm': 1e-6,
        'km' : 1_000,
    }

def foo(data: dict[str, int | float]) -> None: ...

foo(s)
```

For the latest code line I get error message from Pylance:

```
Argument of type "dict[UnitLengthLiterals, int | float]" cannot be assigned to parameter "data" of type "dict[str, int | float]" in function "foo"  "dict[UnitLengthLiterals, int | float]" is incompatible with "dict[str, int | float]"    Type parameter "_KT@dict" is invariant, but "UnitLengthLiterals" is not the same as "str".
```

I have to turn off type checking for this line:

```python
foo(s)  # pyright: ignore[reportGeneralTypeIssues]
```

It seems for me this is a false positive from Pylance/Pyright. What do you think about this?

**Comment by sindzicat** [comment]
I found the answers:

* https://stackoverflow.com/a/73603324
* https://stackoverflow.com/a/73603350
* https://github.com/microsoft/pyright/issues/497

So now I see that it's not Pylance/Pyright bug, and they works correctly in my example.

But I can't use `Mapping` because it's covariant in its second argument, but not in first one as I need.

**Reply by twoertwein** [reply]
You can "abuse" TypeVars for that:

```py
# pyright: strict
from typing import (
    Literal,
    TypeAlias,
    TypeVar,
)

UnitLengthLiterals: TypeAlias = Literal["m", "dm", "cm", "mm", "mkm", "km"]
StrT = TypeVar("StrT", bound=str)

s: dict[UnitLengthLiterals, int | float] = {
    # ...
}

def foo(data: dict[StrT, int | float]) -> None:
    ...

foo(s)
```

**Reply by sindzicat** [reply]
Nice answer, @twoertwein! But I'm not sure, if its' just a hack or a really answer. I'm not so cool in Python typing. But I like your suggestion!

**Reply by twoertwein** [reply]
I'm not sure whether that is documented well (@erictraut) but it is intended https://github.com/microsoft/pyright/issues/2344#issuecomment-927341908

**Comment by sindzicat** [comment]
After reading [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md) I realized why type checker showed an error for me. This is because my function may modify dict so, that its key will not respects to `UnitLengthLiterals`. The example from PEP 586 is type safe, because `str` type is immutable. So my "workaround":

```python
foo(data = {k,v for k,v in s.items()})
```
